{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Isoform protein identification\n",
    " The purpose of this analysis is to extract information on the number of genes with multiple isoforms\n",
    "### Mass-spec information only\n",
    "- Number of genes where isoforms were only identified with shared peptides\n",
    "- Number of genes where a single isoform was identified with unique peptides\n",
    "- Number of genes where multiple isoforms were identified with unique peptides\n",
    "\n",
    "\n",
    "### Transcript incorporation mass spec\n",
    "- Number of genes where a unique isoform was identified with PacBio, but not identified with GENCODE\n",
    "- Number of genes where an isoform that would otherwise be discarded was rescued\n",
    "- Number of genes where ambigous protein groups have multiple confirmed isoforms due to transcript abundance\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from huvec_analysis import data_loader, huvec_config\n",
    "from Bio import SeqIO\n",
    "\n",
    "import os\n",
    "# all statistics go into a directory\n",
    "if not os.path.exists('stats'):\n",
    "    os.makedirs('stats')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "protein_sequences = [] \n",
    "for record in SeqIO.parse(f'{huvec_config.PIPELINE_RESULTS_DIRECTORY}/hybrid_protein_database/huvec_hybrid.fasta', 'fasta'):\n",
    "    acc = record.description.split('|')[1].strip()\n",
    "    gene_name = record.description.split('GN=')[1]\n",
    "    val = {\n",
    "        'accession': acc, \n",
    "        'gene' : gene_name,\n",
    "        'sequence': str(record.seq)\n",
    "    }\n",
    "    protein_sequences.append(val)\n",
    "\n",
    "transcript_abundance = pd.read_table(f'{huvec_config.PIPELINE_RESULTS_DIRECTORY}/hybrid_protein_database/huvec_refined_high_confidence.tsv', usecols=['base_acc', 'CPM'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-3-ff12fd93d46e>\", line 1, in <module>\n",
      "    huvec_peptides = data_loader.read_peptide_file(data_loader.pacbio_hybrid_peptide_file, data_loader.hybrid_gene_map)\n",
      "  File \"/Users/bj8th/Documents/Sheynkman-Lab/GitHub/HUVEC_manuscript_analysis/huvec_analysis/data_loader.py\", line 49, in read_peptide_file\n",
      "    return read_metamorpheus_file(peptides_file, gene_map)\n",
      "  File \"/Users/bj8th/Documents/Sheynkman-Lab/GitHub/HUVEC_manuscript_analysis/huvec_analysis/data_loader.py\", line 41, in read_metamorpheus_file\n",
      "    metamorpheus_table = pd.read_table(metamorpheus_file)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/parsers.py\", line 689, in read_table\n",
      "    return _read(filepath_or_buffer, kwds)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/parsers.py\", line 468, in _read\n",
      "    return parser.read(nrows)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/parsers.py\", line 1057, in read\n",
      "    index, columns, col_dict = self._engine.read(nrows)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/parsers.py\", line 2061, in read\n",
      "    data = self._reader.read(nrows)\n",
      "  File \"pandas/_libs/parsers.pyx\", line 756, in pandas._libs.parsers.TextReader.read\n",
      "  File \"pandas/_libs/parsers.pyx\", line 771, in pandas._libs.parsers.TextReader._read_low_memory\n",
      "  File \"pandas/_libs/parsers.pyx\", line 827, in pandas._libs.parsers.TextReader._read_rows\n",
      "  File \"pandas/_libs/parsers.pyx\", line 814, in pandas._libs.parsers.TextReader._tokenize_rows\n",
      "  File \"pandas/_libs/parsers.pyx\", line 1951, in pandas._libs.parsers.raise_parser_error\n",
      "pandas.errors.ParserError: Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ParserError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 1515, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 1473, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 708, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 737, in getmodule\n",
      "    file = getabsfile(object, _filename)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 721, in getabsfile\n",
      "    return os.path.normcase(os.path.abspath(_filename))\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/posixpath.py\", line 379, in abspath\n",
      "    cwd = os.getcwd()\n",
      "OSError: [Errno 45] Operation not supported\n"
     ]
    }
   ],
   "source": [
    "huvec_peptides = data_loader.read_peptide_file(data_loader.pacbio_hybrid_peptide_file, data_loader.hybrid_gene_map)\n",
    "huvec_peptides['is_high_confidence'] =huvec_peptides['accs'].apply(lambda accs: data_loader.is_high_confidence(accs, data_loader.accs_in_hiconf_space))\n",
    "\n",
    "# huvec_peptides['is_isoform_distinct'] = huvec_peptides['accs'].apply(lambda accs: len(accs) == 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_accessions_and_genes(peptide, protein_sequences):\n",
    "    accessions = []\n",
    "    genes = []\n",
    "    if '|' in peptide:\n",
    "        all_peptides = set(peptide.split('|'))\n",
    "        for val in protein_sequences:\n",
    "            for pep in all_peptides:\n",
    "                if pep in val['sequence']:\n",
    "                    accessions.append(val['accession'])\n",
    "                    genes.append(val['gene'])\n",
    "                    continue\n",
    "    else:\n",
    "        for val in protein_sequences:\n",
    "            if peptide in val['sequence']:\n",
    "                accessions.append(val['accession'])\n",
    "                genes.append(val['gene'])\n",
    "    return [accessions, genes]\n",
    "\n",
    "\n",
    "\n",
    "#huvec_peptides = huvec_peptides.head().copy()\n",
    "huvec_peptides['accessions_genes'] = huvec_peptides['Base Sequence'].apply(lambda peptide: find_accessions_and_genes(peptide, protein_sequences))\n",
    "huvec_peptides['accessions'] = huvec_peptides['accessions_genes'].apply(lambda x : x[0])\n",
    "huvec_peptides['genes'] = huvec_peptides['accessions_genes'].apply(lambda x: x[1])\n",
    "huvec_peptides['num_accessions'] = huvec_peptides['accessions'].apply(len)\n",
    "huvec_peptides['is_peptide_uniquely_mapping'] = huvec_peptides['num_accessions'] == 1 \n",
    "\n",
    "huvec_peptides['is_isoform_distinct'] = huvec_peptides['is_peptide_uniquely_mapping']\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make peptide table\n",
    "Peptide table contains columns in AllPeptides.tsv\n",
    "- filtered to only include target peptides\n",
    "- filtered to only include peptides with QValue < 0.01\n",
    "- map peptide Base Sequence to fasta database\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-6-1695ab7550b2>\", line 9, in <module>\n",
      "    huvec_peptides.to_csv('./stats/huvec_peptides_mapped_accessions.tsv', sep='\\t', index=False)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/core/generic.py\", line 3387, in to_csv\n",
      "    return DataFrameRenderer(formatter).to_csv(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/formats/format.py\", line 1083, in to_csv\n",
      "    csv_formatter.save()\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/formats/csvs.py\", line 228, in save\n",
      "    with get_handle(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/common.py\", line 642, in get_handle\n",
      "    handle = open(\n",
      "TimeoutError: [Errno 60] Operation timed out: './stats/huvec_peptides_mapped_accessions.tsv'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TimeoutError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 1515, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 1473, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 708, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 737, in getmodule\n",
      "    file = getabsfile(object, _filename)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 721, in getabsfile\n",
      "    return os.path.normcase(os.path.abspath(_filename))\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/posixpath.py\", line 379, in abspath\n",
      "    cwd = os.getcwd()\n",
      "TimeoutError: [Errno 60] Operation timed out\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-6-1695ab7550b2>\", line 9, in <module>\n",
      "    huvec_peptides.to_csv('./stats/huvec_peptides_mapped_accessions.tsv', sep='\\t', index=False)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/core/generic.py\", line 3387, in to_csv\n",
      "    return DataFrameRenderer(formatter).to_csv(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/formats/format.py\", line 1083, in to_csv\n",
      "    csv_formatter.save()\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/formats/csvs.py\", line 228, in save\n",
      "    with get_handle(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/common.py\", line 642, in get_handle\n",
      "    handle = open(\n",
      "TimeoutError: [Errno 60] Operation timed out: './stats/huvec_peptides_mapped_accessions.tsv'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TimeoutError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3357, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3454, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2063, in showtraceback\n",
      "    stb = self.InteractiveTB.structured_traceback(etype,\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1367, in structured_traceback\n",
      "    return FormattedTB.structured_traceback(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1267, in structured_traceback\n",
      "    return VerboseTB.structured_traceback(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1124, in structured_traceback\n",
      "    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 1515, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 1473, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 708, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 737, in getmodule\n",
      "    file = getabsfile(object, _filename)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 721, in getabsfile\n",
      "    return os.path.normcase(os.path.abspath(_filename))\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/posixpath.py\", line 379, in abspath\n",
      "    cwd = os.getcwd()\n",
      "TimeoutError: [Errno 60] Operation timed out\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3437, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-6-1695ab7550b2>\", line 9, in <module>\n",
      "    huvec_peptides.to_csv('./stats/huvec_peptides_mapped_accessions.tsv', sep='\\t', index=False)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/core/generic.py\", line 3387, in to_csv\n",
      "    return DataFrameRenderer(formatter).to_csv(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/formats/format.py\", line 1083, in to_csv\n",
      "    csv_formatter.save()\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/formats/csvs.py\", line 228, in save\n",
      "    with get_handle(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/pandas/io/common.py\", line 642, in get_handle\n",
      "    handle = open(\n",
      "TimeoutError: [Errno 60] Operation timed out: './stats/huvec_peptides_mapped_accessions.tsv'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TimeoutError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3357, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3454, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2063, in showtraceback\n",
      "    stb = self.InteractiveTB.structured_traceback(etype,\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1367, in structured_traceback\n",
      "    return FormattedTB.structured_traceback(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1267, in structured_traceback\n",
      "    return VerboseTB.structured_traceback(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1124, in structured_traceback\n",
      "    formatted_exception = self.format_exception_as_a_whole(etype, evalue, etb, number_of_lines_of_context,\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2940, in _run_cell\n",
      "    return runner(coro)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/async_helpers.py\", line 68, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3165, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 3376, in run_ast_nodes\n",
      "    self.showtraceback()\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2063, in showtraceback\n",
      "    stb = self.InteractiveTB.structured_traceback(etype,\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1367, in structured_traceback\n",
      "    return FormattedTB.structured_traceback(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1267, in structured_traceback\n",
      "    return VerboseTB.structured_traceback(\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1142, in structured_traceback\n",
      "    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1082, in format_exception_as_a_whole\n",
      "    last_unique, recursion_repeat = find_recursion(orig_etype, evalue, records)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 382, in find_recursion\n",
      "    return len(records), 0\n",
      "TypeError: object of type 'NoneType' has no len()\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/interactiveshell.py\", line 2061, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 248, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/site-packages/IPython/core/ultratb.py\", line 281, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 1515, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 1473, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 708, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 737, in getmodule\n",
      "    file = getabsfile(object, _filename)\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/inspect.py\", line 721, in getabsfile\n",
      "    return os.path.normcase(os.path.abspath(_filename))\n",
      "  File \"/Users/bj8th/opt/anaconda3/envs/lrpg/lib/python3.8/posixpath.py\", line 379, in abspath\n",
      "    cwd = os.getcwd()\n",
      "TimeoutError: [Errno 60] Operation timed out\n"
     ]
    }
   ],
   "source": [
    "huvec_peptides['accessions'] = huvec_peptides['accessions_genes'].apply(lambda x : x[0])\n",
    "huvec_peptides['genes'] = huvec_peptides['accessions_genes'].apply(lambda x: x[1])\n",
    "huvec_peptides['num_accessions'] = huvec_peptides['accessions'].apply(len)\n",
    "huvec_peptides['is_peptide_uniquely_mapping'] = huvec_peptides['num_accessions'] == 1 \n",
    "\n",
    "huvec_peptides['is_isoform_distinct'] = huvec_peptides['is_peptide_uniquely_mapping']\n",
    "huvec_peptides['accessions'] = huvec_peptides['accessions'].apply(lambda x: '|'.join(x))\n",
    "huvec_peptides['genes'] = huvec_peptides['genes'].apply(lambda x: '|'.join(x))\n",
    "huvec_peptides.to_csv('./stats/huvec_peptides_mapped_accessions.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "huvec_peptides = pd.read_table('./stats/huvec_peptides_mapped_accessions.tsv')\n",
    "huvec_peptides['accessions'] = huvec_peptides['accessions'].apply(lambda x: str(x).split('|'))\n",
    "huvec_peptides['genes'] = huvec_peptides['genes'].apply(lambda x: str(x).split('|'))\n",
    "\n",
    "huvec_peptides['accessions_genes'] = huvec_peptides.apply(lambda row: [row['accessions'], row['genes']] ,axis=1)\n",
    "huvec_peptides['num_accessions'] = huvec_peptides['accessions'].apply(len)\n",
    "huvec_peptides['is_peptide_uniquely_mapping'] = huvec_peptides['num_accessions'] == 1 \n",
    "\n",
    "huvec_peptides['is_isoform_distinct'] = huvec_peptides['is_peptide_uniquely_mapping']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File Name</th>\n",
       "      <th>Scan Number</th>\n",
       "      <th>Scan Retention Time</th>\n",
       "      <th>Num Experimental Peaks</th>\n",
       "      <th>Total Ion Current</th>\n",
       "      <th>Precursor Scan Number</th>\n",
       "      <th>Precursor Charge</th>\n",
       "      <th>Precursor MZ</th>\n",
       "      <th>Precursor Mass</th>\n",
       "      <th>Score</th>\n",
       "      <th>...</th>\n",
       "      <th>PEP</th>\n",
       "      <th>PEP_QValue</th>\n",
       "      <th>accs</th>\n",
       "      <th>genes</th>\n",
       "      <th>is_high_confidence</th>\n",
       "      <th>accessions_genes</th>\n",
       "      <th>accessions</th>\n",
       "      <th>num_accessions</th>\n",
       "      <th>is_peptide_uniquely_mapping</th>\n",
       "      <th>is_isoform_distinct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr11-calib</td>\n",
       "      <td>92300</td>\n",
       "      <td>106.9019</td>\n",
       "      <td>200.0</td>\n",
       "      <td>407915360.0</td>\n",
       "      <td>92257</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1251.59902</td>\n",
       "      <td>3751.77522</td>\n",
       "      <td>41.289</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00004</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>['PB.974.2', 'PB.974.21', 'PB.974.32', 'PB.974...</td>\n",
       "      <td>[LMNA, LMNA, LMNA, LMNA, LMNA]</td>\n",
       "      <td>True</td>\n",
       "      <td>[[PB.974.2, PB.974.21, PB.974.31, PB.974.32, P...</td>\n",
       "      <td>[PB.974.2, PB.974.21, PB.974.31, PB.974.32, PB...</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     File Name  Scan Number  \\\n",
       "2  210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr11-calib        92300   \n",
       "\n",
       "   Scan Retention Time  Num Experimental Peaks  Total Ion Current  \\\n",
       "2             106.9019                   200.0        407915360.0   \n",
       "\n",
       "   Precursor Scan Number  Precursor Charge  Precursor MZ  Precursor Mass  \\\n",
       "2                  92257               3.0    1251.59902      3751.77522   \n",
       "\n",
       "    Score  ...      PEP  PEP_QValue  \\\n",
       "2  41.289  ...  0.00004    0.000028   \n",
       "\n",
       "                                                accs  \\\n",
       "2  ['PB.974.2', 'PB.974.21', 'PB.974.32', 'PB.974...   \n",
       "\n",
       "                            genes is_high_confidence  \\\n",
       "2  [LMNA, LMNA, LMNA, LMNA, LMNA]               True   \n",
       "\n",
       "                                    accessions_genes  \\\n",
       "2  [[PB.974.2, PB.974.21, PB.974.31, PB.974.32, P...   \n",
       "\n",
       "                                          accessions num_accessions  \\\n",
       "2  [PB.974.2, PB.974.21, PB.974.31, PB.974.32, PB...              5   \n",
       "\n",
       "  is_peptide_uniquely_mapping is_isoform_distinct  \n",
       "2                       False               False  \n",
       "\n",
       "[1 rows x 64 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "huvec_peptides = pd.read_table('./stats/huvec_peptides_mapped_accessions.tsv')\n",
    "huvec_peptides['accessions'] = huvec_peptides['accessions'].apply(lambda x: str(x).split('|'))\n",
    "huvec_peptides['genes'] = huvec_peptides['genes'].apply(lambda x: str(x).split('|'))\n",
    "\n",
    "huvec_peptides['accessions_genes'] = huvec_peptides.apply(lambda row: [row['accessions'], row['genes']] ,axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make isoform table\n",
    "The isoform table contains information of the unique and shared peptides identified through mass-spectrometry for each isoform. Also included is the transcript abundance for PacBio isoforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "huvec_peptides['num_isoform_multimap'] = huvec_peptides['accessions'].apply(lambda x: [len(x)] * len(x))\n",
    "huvec_peptides['Sequence'] = huvec_peptides['Base Sequence']\n",
    "exploded = huvec_peptides.set_index(['Sequence'])[['accessions', 'genes', 'num_isoform_multimap']].apply(pd.Series.explode).reset_index()\n",
    "isoform_unique_peptides = exploded[exploded['num_isoform_multimap'] == 1].groupby(['accessions', 'genes'])['Sequence'].apply(list).reset_index()\n",
    "isoform_unique_peptides.rename(columns = {'Sequence' : 'unique_peptides'}, inplace=True)\n",
    "isoform_shared_peptides = exploded[exploded['num_isoform_multimap'] > 1].groupby(['accessions', 'genes'])['Sequence'].apply(list).reset_index()\n",
    "isoform_shared_peptides.rename(columns = {'Sequence' : 'shared_peptides'}, inplace=True)\n",
    "\n",
    "huvec_isoforms = isoform_unique_peptides.merge(isoform_shared_peptides, how = 'outer', on = ['accessions','genes'])\n",
    "huvec_isoforms['is_confirmed_ms'] = ~huvec_isoforms['unique_peptides'].isna()\n",
    "huvec_isoforms.rename(columns={'accessions' : 'accession', 'genes': 'gene'}, inplace=True)\n",
    "huvec_isoforms['unique peptides'] = huvec_isoforms['unique_peptides'].apply(lambda x: '|'.join(x) if type(x) is list else '')\n",
    "huvec_isoforms['shared peptides'] = huvec_isoforms['shared_peptides'].apply(lambda x: '|'.join(x) if type(x) is list else '')\n",
    "huvec_isoforms = huvec_isoforms.merge(transcript_abundance[['base_acc', 'CPM']], left_on='accession', right_on='base_acc', how = 'left')\n",
    "\n",
    "huvec_isoforms.drop(columns=['unique_peptides', 'shared_peptides', 'base_acc'], inplace=True)\n",
    "huvec_isoforms.to_csv('./stats/huvec_isoforms_with_found_peptides.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make gene table\n",
    "The gene table contains information on isoforms confirmed with mass spec as well as isoforms that only contain shared peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "confirmed_isoforms = huvec_isoforms[huvec_isoforms['is_confirmed_ms']][['accession', 'gene']]\n",
    "confirmed_gene_isoforms = confirmed_isoforms.groupby('gene')['accession'].apply(list).reset_index(name = 'isoforms_with_unique_peptides')\n",
    "confirmed_gene_isoforms['number_of_isoforms_with_unique_peptides'] = confirmed_gene_isoforms['isoforms_with_unique_peptides'].apply(len)\n",
    "confirmed_gene_isoforms['isoforms_with_unique_peptides'] = confirmed_gene_isoforms['isoforms_with_unique_peptides'].apply(lambda x: '|'.join(x))\n",
    "\n",
    "shared_isoforms = huvec_isoforms[~huvec_isoforms['is_confirmed_ms']][['accession', 'gene']]\n",
    "shared_gene_isoforms = shared_isoforms.groupby('gene')['accession'].apply(list).reset_index(name = 'isoforms_with_only_shared_peptides')\n",
    "shared_gene_isoforms['number_of_isoforms_with_shared_only_peptides'] = shared_gene_isoforms['isoforms_with_only_shared_peptides'].apply(len)\n",
    "shared_gene_isoforms['isoforms_with_only_shared_peptides'] = shared_gene_isoforms['isoforms_with_only_shared_peptides'].apply(lambda x: '|'.join(x))\n",
    "\n",
    "\n",
    "huvec_genes = pd.merge(confirmed_gene_isoforms, shared_gene_isoforms, on='gene', how = 'outer')\n",
    "huvec_genes['isoforms_with_unique_peptides'].fillna('', inplace=True)\n",
    "huvec_genes['isoforms_with_only_shared_peptides'].fillna('', inplace=True)\n",
    "huvec_genes['number_of_isoforms_with_unique_peptides'].fillna(0, inplace=True)\n",
    "huvec_genes['number_of_isoforms_with_shared_only_peptides'].fillna(0, inplace=True)\n",
    "huvec_genes = huvec_genes.astype({\"number_of_isoforms_with_unique_peptides\":'int', \"number_of_isoforms_with_shared_only_peptides\":'int'}) \n",
    "isoforms_in_database = pd.DataFrame(protein_sequences)\n",
    "genes_database_sizes = isoforms_in_database.groupby('gene').size().reset_index(name='number_isoforms_in_database')\n",
    "huvec_genes = huvec_genes.merge(genes_database_sizes, on = 'gene', how = 'left' )\n",
    "huvec_genes.to_csv('./stats/huvec_genes_found_in_mass_spec.tsv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Breakdown of genes found in mass spec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes found through mass spec\n",
      "10583\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of genes found through mass spec\")\n",
    "print(len(huvec_genes))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All genes: PB high confidence + Gencode\n",
      "The number of genes where there are no unique peptides mapping to any isoform\n",
      "6112\n"
     ]
    }
   ],
   "source": [
    "print('All genes: PB high confidence + Gencode')\n",
    "print('The number of genes where there are no unique peptides mapping to any isoform')\n",
    "print(len(huvec_genes[huvec_genes['isoforms_with_unique_peptides'] == '']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes with at least one unique isoform\n",
      "4471\n",
      "Number of genes with confirmed isoforms where only one isoform is in database\n",
      "1758\n",
      "Number of genes with unique isoforms where  multiple isoforms are in the database\n",
      "2713\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of genes with at least one unique isoform\")\n",
    "print(len(\n",
    "    huvec_genes[\n",
    "        (huvec_genes['number_of_isoforms_with_unique_peptides'] > 0)]\n",
    "    ))\n",
    "print(\"Number of genes with confirmed isoforms where only one isoform is in database\")\n",
    "print(len(\n",
    "    huvec_genes[\n",
    "        (huvec_genes['number_of_isoforms_with_unique_peptides'] > 0) &\n",
    "        (huvec_genes['number_isoforms_in_database'] == 1)]\n",
    "    ))\n",
    "print(\"Number of genes with unique isoforms where  multiple isoforms are in the database\")\n",
    "print(len(\n",
    "    huvec_genes[\n",
    "        (huvec_genes['number_of_isoforms_with_unique_peptides'] > 0) &\n",
    "        (huvec_genes['number_isoforms_in_database']  > 1)]\n",
    "    ))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Breakdown of number of isoforms confirmed per gene\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>number_of_isoforms_with_unique_peptides</th>\n",
       "      <th>number_of_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>6112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>4368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   number_of_isoforms_with_unique_peptides  number_of_genes\n",
       "0                                        0             6112\n",
       "1                                        1             4368\n",
       "2                                        2               92\n",
       "3                                        3               10\n",
       "4                                        7                1"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Breakdown of number of isoforms confirmed per gene\")\n",
    "huvec_genes.groupby('number_of_isoforms_with_unique_peptides').size().reset_index(name='number_of_genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "2f40891b4156b07d6f7078c8cafe72a6146e3c2223ae724d19ed7912498cb5b1"
  },
  "kernelspec": {
   "display_name": "Python 3.8.8 ('lrpg')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
