{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Isoform protein identification\n",
    " The purpose of this analysis is to extract information on the number of genes with multiple isoforms\n",
    "### Mass-spec information only\n",
    "- Number of genes where isoforms were only identified with shared peptides\n",
    "- Number of genes where a single isoform was identified with unique peptides\n",
    "- Number of genes where multiple isoforms were identified with unique peptides\n",
    "\n",
    "\n",
    "### Transcript incorporation mass spec\n",
    "- Number of genes where a unique isoform was identified with PacBio, but not identified with GENCODE\n",
    "- Number of genes where an isoform that would otherwise be discarded was rescued\n",
    "- Number of genes where ambigous protein groups have multiple confirmed isoforms due to transcript abundance\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from huvec_analysis import data_loader, huvec_config\n",
    "from Bio import SeqIO\n",
    "\n",
    "import os\n",
    "# all statistics go into a directory\n",
    "if not os.path.exists('stats'):\n",
    "    os.makedirs('stats')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "protein_sequences = [] \n",
    "for record in SeqIO.parse(f'{huvec_config.PIPELINE_RESULTS_DIRECTORY}/hybrid_protein_database/huvec_hybrid.fasta', 'fasta'):\n",
    "    acc = record.description.split('|')[1].strip()\n",
    "    gene_name = record.description.split('GN=')[1]\n",
    "    val = {\n",
    "        'accession': acc, \n",
    "        'gene' : gene_name,\n",
    "        'sequence': str(record.seq)\n",
    "    }\n",
    "    protein_sequences.append(val)\n",
    "\n",
    "transcript_abundance = pd.read_table(f'{huvec_config.PIPELINE_RESULTS_DIRECTORY}/hybrid_protein_database/huvec_refined_high_confidence.tsv', usecols=['base_acc', 'CPM'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/madison/Documents/Sheynkman_lab/Huvec-project/huvec-proteogenomics/HUVEC_manuscript_analysis/huvec_analysis/data_loader.py:41: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  metamorpheus_table = pd.read_table(metamorpheus_file)\n"
     ]
    }
   ],
   "source": [
    "huvec_peptides = data_loader.read_peptide_file(data_loader.pacbio_hybrid_peptide_file, data_loader.hybrid_gene_map)\n",
    "huvec_peptides['is_high_confidence'] =huvec_peptides['accs'].apply(lambda accs: data_loader.is_high_confidence(accs, data_loader.accs_in_hiconf_space))\n",
    "\n",
    "# huvec_peptides['is_isoform_distinct'] = huvec_peptides['accs'].apply(lambda accs: len(accs) == 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_accessions_and_genes(peptide, protein_sequences):\n",
    "    accessions = []\n",
    "    genes = []\n",
    "    if '|' in peptide:\n",
    "        all_peptides = set(peptide.split('|'))\n",
    "        for val in protein_sequences:\n",
    "            for pep in all_peptides:\n",
    "                if pep in val['sequence']:\n",
    "                    accessions.append(val['accession'])\n",
    "                    genes.append(val['gene'])\n",
    "                    continue\n",
    "    else:\n",
    "        for val in protein_sequences:\n",
    "            if peptide in val['sequence']:\n",
    "                accessions.append(val['accession'])\n",
    "                genes.append(val['gene'])\n",
    "    return [accessions, genes]\n",
    "\n",
    "\n",
    "\n",
    "#huvec_peptides = huvec_peptides.head().copy()\n",
    "huvec_peptides['accessions_genes'] = huvec_peptides['Base Sequence'].apply(lambda peptide: find_accessions_and_genes(peptide, protein_sequences))\n",
    "huvec_peptides['accessions'] = huvec_peptides['accessions_genes'].apply(lambda x : x[0])\n",
    "huvec_peptides['genes'] = huvec_peptides['accessions_genes'].apply(lambda x: x[1])\n",
    "huvec_peptides['num_accessions'] = huvec_peptides['accessions'].apply(len)\n",
    "huvec_peptides['is_peptide_uniquely_mapping'] = huvec_peptides['num_accessions'] == 1 \n",
    "\n",
    "huvec_peptides['is_isoform_distinct'] = huvec_peptides['is_peptide_uniquely_mapping']\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File Name</th>\n",
       "      <th>Scan Number</th>\n",
       "      <th>Scan Retention Time</th>\n",
       "      <th>Num Experimental Peaks</th>\n",
       "      <th>Total Ion Current</th>\n",
       "      <th>Precursor Scan Number</th>\n",
       "      <th>Precursor Charge</th>\n",
       "      <th>Precursor MZ</th>\n",
       "      <th>Precursor Mass</th>\n",
       "      <th>Score</th>\n",
       "      <th>...</th>\n",
       "      <th>Cumulative Decoy</th>\n",
       "      <th>QValue</th>\n",
       "      <th>Cumulative Target Notch</th>\n",
       "      <th>Cumulative Decoy Notch</th>\n",
       "      <th>QValue Notch</th>\n",
       "      <th>PEP</th>\n",
       "      <th>PEP_QValue</th>\n",
       "      <th>accs</th>\n",
       "      <th>genes</th>\n",
       "      <th>is_high_confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr11-calib</td>\n",
       "      <td>184489</td>\n",
       "      <td>189.97150</td>\n",
       "      <td>200.0</td>\n",
       "      <td>5.592743e+07</td>\n",
       "      <td>184417</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1240.94293</td>\n",
       "      <td>3719.80695</td>\n",
       "      <td>44.234</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>[SOD1-201]</td>\n",
       "      <td>[SOD1]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr6-calib</td>\n",
       "      <td>100665</td>\n",
       "      <td>111.16888</td>\n",
       "      <td>200.0</td>\n",
       "      <td>7.441615e+06</td>\n",
       "      <td>100651</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1118.20901</td>\n",
       "      <td>3351.60521</td>\n",
       "      <td>42.334</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>[PB.1380.5, PB.1380.6]</td>\n",
       "      <td>[CEP170, CEP170]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr11-calib</td>\n",
       "      <td>92300</td>\n",
       "      <td>106.90190</td>\n",
       "      <td>200.0</td>\n",
       "      <td>4.079154e+08</td>\n",
       "      <td>92257</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1251.59902</td>\n",
       "      <td>3751.77522</td>\n",
       "      <td>41.289</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>[PB.974.2, PB.974.21, PB.974.32, PB.974.6]</td>\n",
       "      <td>[LMNA, LMNA, LMNA, LMNA]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr12-calib</td>\n",
       "      <td>202802</td>\n",
       "      <td>200.29734</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1.130036e+08</td>\n",
       "      <td>202737</td>\n",
       "      <td>4.0</td>\n",
       "      <td>981.50912</td>\n",
       "      <td>3922.00739</td>\n",
       "      <td>41.222</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>[PB.6906.1, PB.6906.20, PB.6906.5]</td>\n",
       "      <td>[VIM, VIM, VIM]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr12-calib</td>\n",
       "      <td>186991</td>\n",
       "      <td>188.33567</td>\n",
       "      <td>200.0</td>\n",
       "      <td>3.161935e+07</td>\n",
       "      <td>186955</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1351.04166</td>\n",
       "      <td>4050.10314</td>\n",
       "      <td>41.215</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>[PB.6906.1, PB.6906.20, PB.6906.5]</td>\n",
       "      <td>[VIM, VIM, VIM]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102958</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr16-calib</td>\n",
       "      <td>86672</td>\n",
       "      <td>120.70787</td>\n",
       "      <td>115.0</td>\n",
       "      <td>2.931054e+06</td>\n",
       "      <td>86635</td>\n",
       "      <td>4.0</td>\n",
       "      <td>559.03554</td>\n",
       "      <td>2232.11305</td>\n",
       "      <td>8.089</td>\n",
       "      <td>...</td>\n",
       "      <td>1018</td>\n",
       "      <td>0.009986</td>\n",
       "      <td>86319</td>\n",
       "      <td>788</td>\n",
       "      <td>0.009129</td>\n",
       "      <td>0.723371</td>\n",
       "      <td>0.052111</td>\n",
       "      <td>[ANKFY1-201, ANKFY1-202, ANKFY1-209]</td>\n",
       "      <td>[ANKFY1, ANKFY1, ANKFY1]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102959</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr14-calib</td>\n",
       "      <td>58986</td>\n",
       "      <td>75.21752</td>\n",
       "      <td>167.0</td>\n",
       "      <td>6.655044e+06</td>\n",
       "      <td>58962</td>\n",
       "      <td>2.0</td>\n",
       "      <td>386.74999</td>\n",
       "      <td>771.48542</td>\n",
       "      <td>8.089</td>\n",
       "      <td>...</td>\n",
       "      <td>1018</td>\n",
       "      <td>0.009986</td>\n",
       "      <td>86320</td>\n",
       "      <td>788</td>\n",
       "      <td>0.009129</td>\n",
       "      <td>0.016489</td>\n",
       "      <td>0.001045</td>\n",
       "      <td>[PB.7509.2]</td>\n",
       "      <td>[GTF2H1]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102961</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr13-calib</td>\n",
       "      <td>43102</td>\n",
       "      <td>58.88954</td>\n",
       "      <td>91.0</td>\n",
       "      <td>2.964866e+06</td>\n",
       "      <td>43084</td>\n",
       "      <td>3.0</td>\n",
       "      <td>412.22734</td>\n",
       "      <td>1233.66019</td>\n",
       "      <td>8.088</td>\n",
       "      <td>...</td>\n",
       "      <td>1019</td>\n",
       "      <td>0.009996</td>\n",
       "      <td>15623</td>\n",
       "      <td>230</td>\n",
       "      <td>0.014719</td>\n",
       "      <td>0.000972</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>[PB.11098.3]</td>\n",
       "      <td>[DHX8]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102962</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr9-calib</td>\n",
       "      <td>124111</td>\n",
       "      <td>148.08855</td>\n",
       "      <td>200.0</td>\n",
       "      <td>1.111659e+07</td>\n",
       "      <td>124075</td>\n",
       "      <td>3.0</td>\n",
       "      <td>709.33997</td>\n",
       "      <td>2124.99807</td>\n",
       "      <td>8.088</td>\n",
       "      <td>...</td>\n",
       "      <td>1019</td>\n",
       "      <td>0.009996</td>\n",
       "      <td>15624</td>\n",
       "      <td>230</td>\n",
       "      <td>0.014719</td>\n",
       "      <td>0.991053</td>\n",
       "      <td>0.376097</td>\n",
       "      <td>[PB.7914.4, PB.7914.5]</td>\n",
       "      <td>[USP35, USP35]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102963</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr14-calib</td>\n",
       "      <td>20746</td>\n",
       "      <td>32.48679</td>\n",
       "      <td>171.0</td>\n",
       "      <td>8.594228e+05</td>\n",
       "      <td>20735</td>\n",
       "      <td>3.0</td>\n",
       "      <td>383.54948</td>\n",
       "      <td>1147.62662</td>\n",
       "      <td>8.088</td>\n",
       "      <td>...</td>\n",
       "      <td>1019</td>\n",
       "      <td>0.009996</td>\n",
       "      <td>86321</td>\n",
       "      <td>789</td>\n",
       "      <td>0.009140</td>\n",
       "      <td>0.000618</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>[PB.6339.16]</td>\n",
       "      <td>[UBAP2]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101945 rows × 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          File Name  Scan Number  \\\n",
       "0       210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr11-calib       184489   \n",
       "1        210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr6-calib       100665   \n",
       "2       210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr11-calib        92300   \n",
       "3       210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr12-calib       202802   \n",
       "4       210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr12-calib       186991   \n",
       "...                                             ...          ...   \n",
       "102958  210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr16-calib        86672   \n",
       "102959  210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr14-calib        58986   \n",
       "102961  210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr13-calib        43102   \n",
       "102962   210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr9-calib       124111   \n",
       "102963  210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr14-calib        20746   \n",
       "\n",
       "        Scan Retention Time  Num Experimental Peaks  Total Ion Current  \\\n",
       "0                 189.97150                   200.0       5.592743e+07   \n",
       "1                 111.16888                   200.0       7.441615e+06   \n",
       "2                 106.90190                   200.0       4.079154e+08   \n",
       "3                 200.29734                   200.0       1.130036e+08   \n",
       "4                 188.33567                   200.0       3.161935e+07   \n",
       "...                     ...                     ...                ...   \n",
       "102958            120.70787                   115.0       2.931054e+06   \n",
       "102959             75.21752                   167.0       6.655044e+06   \n",
       "102961             58.88954                    91.0       2.964866e+06   \n",
       "102962            148.08855                   200.0       1.111659e+07   \n",
       "102963             32.48679                   171.0       8.594228e+05   \n",
       "\n",
       "        Precursor Scan Number  Precursor Charge  Precursor MZ  Precursor Mass  \\\n",
       "0                      184417               3.0    1240.94293      3719.80695   \n",
       "1                      100651               3.0    1118.20901      3351.60521   \n",
       "2                       92257               3.0    1251.59902      3751.77522   \n",
       "3                      202737               4.0     981.50912      3922.00739   \n",
       "4                      186955               3.0    1351.04166      4050.10314   \n",
       "...                       ...               ...           ...             ...   \n",
       "102958                  86635               4.0     559.03554      2232.11305   \n",
       "102959                  58962               2.0     386.74999       771.48542   \n",
       "102961                  43084               3.0     412.22734      1233.66019   \n",
       "102962                 124075               3.0     709.33997      2124.99807   \n",
       "102963                  20735               3.0     383.54948      1147.62662   \n",
       "\n",
       "         Score  ...  Cumulative Decoy    QValue Cumulative Target Notch  \\\n",
       "0       44.234  ...                 0  0.000000                       1   \n",
       "1       42.334  ...                 0  0.000000                       1   \n",
       "2       41.289  ...                 0  0.000000                       2   \n",
       "3       41.222  ...                 0  0.000000                       3   \n",
       "4       41.215  ...                 0  0.000000                       4   \n",
       "...        ...  ...               ...       ...                     ...   \n",
       "102958   8.089  ...              1018  0.009986                   86319   \n",
       "102959   8.089  ...              1018  0.009986                   86320   \n",
       "102961   8.088  ...              1019  0.009996                   15623   \n",
       "102962   8.088  ...              1019  0.009996                   15624   \n",
       "102963   8.088  ...              1019  0.009996                   86321   \n",
       "\n",
       "       Cumulative Decoy Notch QValue Notch       PEP  PEP_QValue  \\\n",
       "0                           0     0.000000  0.000024    0.000016   \n",
       "1                           0     0.000000  0.000085    0.000059   \n",
       "2                           0     0.000000  0.000040    0.000028   \n",
       "3                           0     0.000000  0.000028    0.000019   \n",
       "4                           0     0.000000  0.000018    0.000012   \n",
       "...                       ...          ...       ...         ...   \n",
       "102958                    788     0.009129  0.723371    0.052111   \n",
       "102959                    788     0.009129  0.016489    0.001045   \n",
       "102961                    230     0.014719  0.000972    0.000168   \n",
       "102962                    230     0.014719  0.991053    0.376097   \n",
       "102963                    789     0.009140  0.000618    0.000142   \n",
       "\n",
       "                                              accs                     genes  \\\n",
       "0                                       [SOD1-201]                    [SOD1]   \n",
       "1                           [PB.1380.5, PB.1380.6]          [CEP170, CEP170]   \n",
       "2       [PB.974.2, PB.974.21, PB.974.32, PB.974.6]  [LMNA, LMNA, LMNA, LMNA]   \n",
       "3               [PB.6906.1, PB.6906.20, PB.6906.5]           [VIM, VIM, VIM]   \n",
       "4               [PB.6906.1, PB.6906.20, PB.6906.5]           [VIM, VIM, VIM]   \n",
       "...                                            ...                       ...   \n",
       "102958        [ANKFY1-201, ANKFY1-202, ANKFY1-209]  [ANKFY1, ANKFY1, ANKFY1]   \n",
       "102959                                 [PB.7509.2]                  [GTF2H1]   \n",
       "102961                                [PB.11098.3]                    [DHX8]   \n",
       "102962                      [PB.7914.4, PB.7914.5]            [USP35, USP35]   \n",
       "102963                                [PB.6339.16]                   [UBAP2]   \n",
       "\n",
       "       is_high_confidence  \n",
       "0                   False  \n",
       "1                    True  \n",
       "2                    True  \n",
       "3                    True  \n",
       "4                    True  \n",
       "...                   ...  \n",
       "102958              False  \n",
       "102959               True  \n",
       "102961               True  \n",
       "102962               True  \n",
       "102963               True  \n",
       "\n",
       "[101945 rows x 59 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "huvec_peptides"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make peptide table\n",
    "Peptide table contains columns in AllPeptides.tsv\n",
    "- filtered to only include target peptides\n",
    "- filtered to only include peptides with QValue < 0.01\n",
    "- map peptide Base Sequence to fasta database\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "huvec_peptides['accessions'] = huvec_peptides['accessions_genes'].apply(lambda x : x[0])\n",
    "huvec_peptides['genes'] = huvec_peptides['accessions_genes'].apply(lambda x: x[1])\n",
    "huvec_peptides['num_accessions'] = huvec_peptides['accessions'].apply(len)\n",
    "huvec_peptides['is_peptide_uniquely_mapping'] = huvec_peptides['num_accessions'] == 1 \n",
    "\n",
    "huvec_peptides['is_isoform_distinct'] = huvec_peptides['is_peptide_uniquely_mapping']\n",
    "huvec_peptides['accessions'] = huvec_peptides['accessions'].apply(lambda x: '|'.join(x))\n",
    "huvec_peptides['genes'] = huvec_peptides['genes'].apply(lambda x: '|'.join(x))\n",
    "huvec_peptides.to_csv('./stats/huvec_peptides_mapped_accessions.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "huvec_peptides = pd.read_table('./stats/huvec_peptides_mapped_accessions.tsv')\n",
    "huvec_peptides['accessions'] = huvec_peptides['accessions'].apply(lambda x: str(x).split('|'))\n",
    "huvec_peptides['genes'] = huvec_peptides['genes'].apply(lambda x: str(x).split('|'))\n",
    "\n",
    "huvec_peptides['accessions_genes'] = huvec_peptides.apply(lambda row: [row['accessions'], row['genes']] ,axis=1)\n",
    "huvec_peptides['num_accessions'] = huvec_peptides['accessions'].apply(len)\n",
    "huvec_peptides['is_peptide_uniquely_mapping'] = huvec_peptides['num_accessions'] == 1 \n",
    "\n",
    "huvec_peptides['is_isoform_distinct'] = huvec_peptides['is_peptide_uniquely_mapping']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File Name</th>\n",
       "      <th>Scan Number</th>\n",
       "      <th>Scan Retention Time</th>\n",
       "      <th>Num Experimental Peaks</th>\n",
       "      <th>Total Ion Current</th>\n",
       "      <th>Precursor Scan Number</th>\n",
       "      <th>Precursor Charge</th>\n",
       "      <th>Precursor MZ</th>\n",
       "      <th>Precursor Mass</th>\n",
       "      <th>Score</th>\n",
       "      <th>...</th>\n",
       "      <th>PEP</th>\n",
       "      <th>PEP_QValue</th>\n",
       "      <th>accs</th>\n",
       "      <th>genes</th>\n",
       "      <th>is_high_confidence</th>\n",
       "      <th>accessions_genes</th>\n",
       "      <th>accessions</th>\n",
       "      <th>num_accessions</th>\n",
       "      <th>is_peptide_uniquely_mapping</th>\n",
       "      <th>is_isoform_distinct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr11-calib</td>\n",
       "      <td>92300</td>\n",
       "      <td>106.9019</td>\n",
       "      <td>200.0</td>\n",
       "      <td>407915360.0</td>\n",
       "      <td>92257</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1251.59902</td>\n",
       "      <td>3751.77522</td>\n",
       "      <td>41.289</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00004</td>\n",
       "      <td>0.000028</td>\n",
       "      <td>['PB.974.2', 'PB.974.21', 'PB.974.32', 'PB.974...</td>\n",
       "      <td>[LMNA, LMNA, LMNA, LMNA, LMNA]</td>\n",
       "      <td>True</td>\n",
       "      <td>[[PB.974.2, PB.974.21, PB.974.31, PB.974.32, P...</td>\n",
       "      <td>[PB.974.2, PB.974.21, PB.974.31, PB.974.32, PB...</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 64 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     File Name  Scan Number  \\\n",
       "2  210827_HUVEC5,6_tryp_HCDorbi_4hr_Fr11-calib        92300   \n",
       "\n",
       "   Scan Retention Time  Num Experimental Peaks  Total Ion Current  \\\n",
       "2             106.9019                   200.0        407915360.0   \n",
       "\n",
       "   Precursor Scan Number  Precursor Charge  Precursor MZ  Precursor Mass  \\\n",
       "2                  92257               3.0    1251.59902      3751.77522   \n",
       "\n",
       "    Score  ...      PEP  PEP_QValue  \\\n",
       "2  41.289  ...  0.00004    0.000028   \n",
       "\n",
       "                                                accs  \\\n",
       "2  ['PB.974.2', 'PB.974.21', 'PB.974.32', 'PB.974...   \n",
       "\n",
       "                            genes is_high_confidence  \\\n",
       "2  [LMNA, LMNA, LMNA, LMNA, LMNA]               True   \n",
       "\n",
       "                                    accessions_genes  \\\n",
       "2  [[PB.974.2, PB.974.21, PB.974.31, PB.974.32, P...   \n",
       "\n",
       "                                          accessions num_accessions  \\\n",
       "2  [PB.974.2, PB.974.21, PB.974.31, PB.974.32, PB...              5   \n",
       "\n",
       "  is_peptide_uniquely_mapping is_isoform_distinct  \n",
       "2                       False               False  \n",
       "\n",
       "[1 rows x 64 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "huvec_peptides = pd.read_table('./stats/huvec_peptides_mapped_accessions.tsv')\n",
    "huvec_peptides['accessions'] = huvec_peptides['accessions'].apply(lambda x: str(x).split('|'))\n",
    "huvec_peptides['genes'] = huvec_peptides['genes'].apply(lambda x: str(x).split('|'))\n",
    "\n",
    "huvec_peptides['accessions_genes'] = huvec_peptides.apply(lambda row: [row['accessions'], row['genes']] ,axis=1)\n",
    "\n",
    "huvec_peptides[huvec_peptides['Base Sequence'] == 'TVLCGTCGQPADKASASGSGAQVGGPISSGSSASSVTVTR']\n",
    "huvec_peptides.at[2, 'accessions_genes'] = find_accessions_and_genes('TVLCGTCGQPADKASASGSGAQVGGPISSGSSASSVTVTR', protein_sequences)\n",
    "huvec_peptides[huvec_peptides['Base Sequence'] == 'TVLCGTCGQPADKASASGSGAQVGGPISSGSSASSVTVTR']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make isoform table\n",
    "The isoform table contains information of the unique and shared peptides identified through mass-spectrometry for each isoform. Also included is the transcript abundance for PacBio isoforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "huvec_peptides['num_isoform_multimap'] = huvec_peptides['accessions'].apply(lambda x: [len(x)] * len(x))\n",
    "huvec_peptides['Sequence'] = huvec_peptides['Base Sequence']\n",
    "exploded = huvec_peptides.set_index(['Sequence'])[['accessions', 'genes', 'num_isoform_multimap']].apply(pd.Series.explode).reset_index()\n",
    "isoform_unique_peptides = exploded[exploded['num_isoform_multimap'] == 1].groupby(['accessions', 'genes'])['Sequence'].apply(list).reset_index()\n",
    "isoform_unique_peptides.rename(columns = {'Sequence' : 'unique_peptides'}, inplace=True)\n",
    "isoform_shared_peptides = exploded[exploded['num_isoform_multimap'] > 1].groupby(['accessions', 'genes'])['Sequence'].apply(list).reset_index()\n",
    "isoform_shared_peptides.rename(columns = {'Sequence' : 'shared_peptides'}, inplace=True)\n",
    "\n",
    "huvec_isoforms = isoform_unique_peptides.merge(isoform_shared_peptides, how = 'outer', on = ['accessions','genes'])\n",
    "huvec_isoforms['is_confirmed_ms'] = ~huvec_isoforms['unique_peptides'].isna()\n",
    "huvec_isoforms.rename(columns={'accessions' : 'accession', 'genes': 'gene'}, inplace=True)\n",
    "huvec_isoforms['unique peptides'] = huvec_isoforms['unique_peptides'].apply(lambda x: '|'.join(x) if type(x) is list else '')\n",
    "huvec_isoforms['shared peptides'] = huvec_isoforms['shared_peptides'].apply(lambda x: '|'.join(x) if type(x) is list else '')\n",
    "huvec_isoforms = huvec_isoforms.merge(transcript_abundance[['base_acc', 'CPM']], left_on='accession', right_on='base_acc', how = 'left')\n",
    "\n",
    "huvec_isoforms.drop(columns=['unique_peptides', 'shared_peptides', 'base_acc'], inplace=True)\n",
    "huvec_isoforms.to_csv('./stats/huvec_isoforms_with_found_peptides.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "huvec_isoforms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make gene table\n",
    "The gene table contains information on isoforms confirmed with mass spec as well as isoforms that only contain shared peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "confirmed_isoforms = huvec_isoforms[huvec_isoforms['is_confirmed_ms']][['accession', 'gene']]\n",
    "confirmed_gene_isoforms = confirmed_isoforms.groupby('gene')['accession'].apply(list).reset_index(name = 'isoforms_with_unique_peptides')\n",
    "confirmed_gene_isoforms['number_of_isoforms_with_unique_peptides'] = confirmed_gene_isoforms['isoforms_with_unique_peptides'].apply(len)\n",
    "confirmed_gene_isoforms['isoforms_with_unique_peptides'] = confirmed_gene_isoforms['isoforms_with_unique_peptides'].apply(lambda x: '|'.join(x))\n",
    "\n",
    "shared_isoforms = huvec_isoforms[~huvec_isoforms['is_confirmed_ms']][['accession', 'gene']]\n",
    "shared_gene_isoforms = shared_isoforms.groupby('gene')['accession'].apply(list).reset_index(name = 'isoforms_with_only_shared_peptides')\n",
    "shared_gene_isoforms['number_of_isoforms_with_shared_only_peptides'] = shared_gene_isoforms['isoforms_with_only_shared_peptides'].apply(len)\n",
    "shared_gene_isoforms['isoforms_with_only_shared_peptides'] = shared_gene_isoforms['isoforms_with_only_shared_peptides'].apply(lambda x: '|'.join(x))\n",
    "\n",
    "\n",
    "huvec_genes = pd.merge(confirmed_gene_isoforms, shared_gene_isoforms, on='gene', how = 'outer')\n",
    "huvec_genes['isoforms_with_unique_peptides'].fillna('', inplace=True)\n",
    "huvec_genes['isoforms_with_only_shared_peptides'].fillna('', inplace=True)\n",
    "huvec_genes['number_of_isoforms_with_unique_peptides'].fillna(0, inplace=True)\n",
    "huvec_genes['number_of_isoforms_with_shared_only_peptides'].fillna(0, inplace=True)\n",
    "huvec_genes = huvec_genes.astype({\"number_of_isoforms_with_unique_peptides\":'int', \"number_of_isoforms_with_shared_only_peptides\":'int'}) \n",
    "isoforms_in_database = pd.DataFrame(protein_sequences)\n",
    "genes_database_sizes = isoforms_in_database.groupby('gene').size().reset_index(name='number_isoforms_in_database')\n",
    "huvec_genes = huvec_genes.merge(genes_database_sizes, on = 'gene', how = 'left' )\n",
    "huvec_genes.to_csv('./220331_stats/huvec_genes_found_in_mass_spec.tsv', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Breakdown of genes found in mass spec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes found through mass spec\n",
      "10583\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of genes found through mass spec\")\n",
    "print(len(huvec_genes))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All genes: PB high confidence + Gencode\n",
      "The number of genes where there are no unique peptides mapping to any isoform\n",
      "6112\n"
     ]
    }
   ],
   "source": [
    "print('All genes: PB high confidence + Gencode')\n",
    "print('The number of genes where there are no unique peptides mapping to any isoform')\n",
    "print(len(huvec_genes[huvec_genes['isoforms_with_unique_peptides'] == '']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes with at least one unique isoform\n",
      "4471\n",
      "Number of genes with confirmed isoforms where only one isoform is in database\n",
      "1758\n",
      "Number of genes with unique isoforms where  multiple isoforms are in the database\n",
      "2713\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of genes with at least one unique isoform\")\n",
    "print(len(\n",
    "    huvec_genes[\n",
    "        (huvec_genes['number_of_isoforms_with_unique_peptides'] > 0)]\n",
    "    ))\n",
    "print(\"Number of genes with confirmed isoforms where only one isoform is in database\")\n",
    "print(len(\n",
    "    huvec_genes[\n",
    "        (huvec_genes['number_of_isoforms_with_unique_peptides'] > 0) &\n",
    "        (huvec_genes['number_isoforms_in_database'] == 1)]\n",
    "    ))\n",
    "print(\"Number of genes with unique isoforms where  multiple isoforms are in the database\")\n",
    "print(len(\n",
    "    huvec_genes[\n",
    "        (huvec_genes['number_of_isoforms_with_unique_peptides'] > 0) &\n",
    "        (huvec_genes['number_isoforms_in_database']  > 1)]\n",
    "    ))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Breakdown of number of isoforms confirmed per gene\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>number_of_isoforms_with_unique_peptides</th>\n",
       "      <th>number_of_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>6112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>4368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   number_of_isoforms_with_unique_peptides  number_of_genes\n",
       "0                                        0             6112\n",
       "1                                        1             4368\n",
       "2                                        2               92\n",
       "3                                        3               10\n",
       "4                                        7                1"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Breakdown of number of isoforms confirmed per gene\")\n",
    "huvec_genes.groupby('number_of_isoforms_with_unique_peptides').size().reset_index(name='number_of_genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "2f40891b4156b07d6f7078c8cafe72a6146e3c2223ae724d19ed7912498cb5b1"
  },
  "kernelspec": {
   "display_name": "Python 3.8.8 ('lrpg')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
