{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Isoform Peptide multi-mapping \n",
    "file in order to look thru the AllPeptides file \n",
    "- find the genes and isoforms and the peptided\n",
    "- find if the gene is multi-mapping or not "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from huvec_analysis import huvec_config\n",
    "# from huvec_analysis.data_loader import genes_in_hiconf_space, accs_in_hiconf_space\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "from Bio import SeqIO\n",
    "import os \n",
    "import matplotlib\n",
    "\n",
    "matplotlib.rc('font', **huvec_config.font)\n",
    "\n",
    "# all plots go into a directory\n",
    "if not os.path.exists('plot'):\n",
    "    os.makedirs('plot')\n",
    "\n",
    "# all statistics go into a directory\n",
    "if not os.path.exists('stats'):\n",
    "    os.makedirs('stats')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#read in the file paths \n",
    "#toy_fasta_file = f'{huvec_config.REFERENCE_DIRECTORY}/huvec_toy.fasta'\n",
    "fasta_file = f'{huvec_config.PIPELINE_RESULTS_DIRECTORY}/hybrid_protein_database/huvec_hybrid.fasta'\n",
    "pacbio_protein_grps_path = '../00_pre_analysis/metamorpheus_table/AllProteinGroups.PacBioHybrid.tsv'\n",
    "pacbio_peps_path = f'{huvec_config.PIPELINE_RESULTS_DIRECTORY}/metamorpheus/pacbio/hybrid/search_results/Task1SearchTask/AllPeptides.huvec.hybrid.psmtsv'\n",
    "pacbio_peps_filtered_path = '../00_pre_analysis/metamorpheus_table/AllPeptides.PacBioHybrid.tsv'\n",
    "sqanti_info_path = '../00_pre_analysis/sqanti_info/sqanti_info.tsv'\n",
    "# make a toy file of the peps file \n",
    "# read in the peps file \n",
    "pacbio_peps_filtered = pd.read_table(pacbio_peps_filtered_path)\n",
    "\n",
    "#read in the sqanti table for com information \n",
    "sqanti_info = pd.read_table(sqanti_info_path)\n",
    "#%%\n",
    "# Gene|isoform|has_unique_peptide (yes.no)|listing of the unique peptides \n",
    "# read in the fasta file with the biopython module \n",
    "##test whih part of the sequence we are extracting \n",
    "\n",
    "fasta_sequences_pb = SeqIO.parse(fasta_file, 'fasta')\n",
    "\n",
    "def find_trans_matches(peptide_seq, fasta_seq):\n",
    "    '''\n",
    "    Reading through the fasta sequence find if there is a matching \n",
    "    peptide within the fasta sequence\n",
    "    if so then the transcript id of the match will be returned'''\n",
    "    transcript_ids = []\n",
    "    for trans_id, sequence in fasta_seq.items(): \n",
    "       # print (trans_id, sequence)\n",
    "        \n",
    "        if peptide_seq in sequence:\n",
    "            transcript_ids.append(trans_id)\n",
    "        #break\n",
    "    return transcript_ids\n",
    "    #return seq.description\n",
    "\n",
    "#create dict\n",
    "#create dictionary for the pacbio sequences\n",
    "def process_fasta_file_pb(fasta_file_pb):\n",
    "    \"\"\"\n",
    "    For configuration, to convert the global fasta sequence file into a \n",
    "    sequence dictionary that can be used in mapping\n",
    "    Parameters\n",
    "    ---------\n",
    "    fasta_file : str\n",
    "        file location of fasta file\n",
    "    Returns\n",
    "    ---------\n",
    "    sequences : dict\n",
    "        {acc : sequence} dictionary generated from fasta file\n",
    "    \"\"\"\n",
    "    seqs = SeqIO.parse(open(fasta_file_pb), 'fasta')\n",
    "    sequences = {}\n",
    "    for entry in seqs:\n",
    "        seq = str(entry.seq)\n",
    "        #acc = entry.description.split('|')[1].strip()\n",
    "        acc = entry.description\n",
    "        sequences[acc] = seq\n",
    "    return sequences\n",
    "\n",
    "#apply the function to the pb fata file \n",
    "fasta_sequences_pb = process_fasta_file_pb(fasta_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#apply the functions and add to the peps table to find what transcripts we have from pacbio present within gencode\n",
    "# Perform for the entire length of the rows for AllPeptides file\n",
    "pacbio_peps_filtered['transcripts_pb']=pacbio_peps_filtered.apply(lambda row:find_trans_matches(row['Base Sequence'],fasta_sequences_pb),axis=1)\n",
    "\n",
    "# convert the list to a string for searching downstream\n",
    "pacbio_peps_filtered['transcripts_pb_str'] = pacbio_peps_filtered['transcripts_pb'].astype(str)\n",
    "#find the rows that have more than 1 accession listed (based on the comma)\n",
    "\n",
    "# find where the length of the list is greater than 1 (meaning there is is not unique)\n",
    "def has_multiple(group):\n",
    "    return len(group) <=1\n",
    "pacbio_peps_filtered['has_one_iso'] = pacbio_peps_filtered['transcripts_pb'].apply(has_multiple)\n",
    "\n",
    "# converts the True false to 1 for yes 0 for no\n",
    "pacbio_peps_filtered['peptide_is_unique'] = pacbio_peps_filtered['has_one_iso'].astype(int)\n",
    "pacbio_peps_filtered.to_csv('stats/peptide_isoform_coverage.tsv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "unique_isos = pacbio_peps_filtered[pacbio_peps_filtered['peptide_is_unique']==1]\n",
    "unique_isos = unique_isos[['gene_name', 'Protein Accession', 'peptide_is_unique',  'ec_priority', 'Base Sequence']]\n",
    "\n",
    "# find the unique EC peptides\n",
    "ec_unique_isos = unique_isos[unique_isos['ec_priority']==1]\n",
    "ec_unique_isos.to_csv('stats/ec_genes_unique_isos.tsv', sep = '\\t')\n",
    "\n",
    "# find the number of multi-mapping isoforms\n",
    "not_unique_isos = pacbio_peps_filtered[pacbio_peps_filtered['peptide_is_unique']==0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    " # find for how many isoforms we have exactly 2 unique accessions (w/ ec gene column)\n",
    "gene_multiple_w_ec_genes = unique_isos[['gene_name', 'Protein Accession', 'ec_priority']].drop_duplicates()\n",
    "gene_multiple_w_ec_genes = gene_multiple_w_ec_genes.groupby(['gene_name', 'ec_priority']).size().reset_index(name='size')\n",
    "one_uniq_iso = gene_multiple_w_ec_genes[gene_multiple_w_ec_genes['size'] == 1]\n",
    "gene_multiple_w_ec_genes = gene_multiple_w_ec_genes[gene_multiple_w_ec_genes['size'] > 1]\n",
    "gene_multiple_w_ec_genes.to_csv('stats/genes_w_multiple_w_ec_genes.tsv', sep = '\\t', index = None)\n",
    "\n",
    "# find for how many genes there are exactly 2 isoforms for a gene with unique peptide evidence \n",
    "two_uniq_isos = gene_multiple_w_ec_genes[gene_multiple_w_ec_genes['size']==2]\n",
    "\n",
    "ec_genes_two_uniq_isos = gene_multiple_w_ec_genes[(gene_multiple_w_ec_genes['size']==2) & (gene_multiple_w_ec_genes['ec_priority']==1)]\n",
    "\n",
    "three_uniq_isos = gene_multiple_w_ec_genes[gene_multiple_w_ec_genes['size']==3]\n",
    "\n",
    "four_uniq_isos = gene_multiple_w_ec_genes[gene_multiple_w_ec_genes['size']==4]\n",
    "\n",
    "# merge the sqanti info with the peptides table for cpm information \n",
    "pacbio_peps_filtered['pr_acc'] = pacbio_peps_filtered['Protein Accession'].str.split('|')\n",
    "\n",
    "# supplementary table outp\n",
    "supp_table_isoform_informative = pacbio_peps_filtered[['gene_name', 'Protein Accession', 'peptide_is_unique']]\n",
    "supp_table_isoform_informative.to_csv('stats/supp_table_isoform_informative_table.tsv', sep = '\\t')"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "2f40891b4156b07d6f7078c8cafe72a6146e3c2223ae724d19ed7912498cb5b1"
  },
  "kernelspec": {
   "display_name": "Python 3.8.8 ('lrpg')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
